File: solutions/go/blackjack/1/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
	case "ace":
		return 11
	case "two":
		return 2
	case "three":
		return 3
	case "four":
		return 4
	case "five":
		return 5
	case "six":
		return 6
	case "seven":
		return 7
	case "eight":
		return 8
	case "nine":
		return 9
	case "ten", "jack", "king", "queen":
		return 10
	default:
		return 0
	}
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	total := ParseCard(card1) + ParseCard(card2)
	dealerValue := ParseCard(dealerCard)
	switch {
	case card1 == "ace" && card2 == "ace":
		return "P"
	case (total) == 21:
		if dealerValue < 10 {
			return "W"
		} else {
			return "S"
		}
	case total >= 17 && total <= 20:
		return "S"
	case total >= 12 && total <= 16:
		if dealerValue >= 7 {
			return "H"
		} else {
			return "S"
		}
	case total <= 11:
		return "H"
	default:
		return "W"
	}
}

File: solutions/ruby/blackjack/1/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
      when "ace" then 11
      when "two" then 2
      when "three" then 3
      when "four" then 4
      when "five" then 5
      when "six" then 6
      when "seven" then 7
      when "eight" then 8
      when "nine" then 9
      when "ten", "jack", "queen", "king" then 10
    else 0
    end
  end

  def self.card_range(card1, card2)
    sum = self.parse_card(card1) + self.parse_card(card2)
    case sum
      when 4..11 then "low"
      when 12..16 then "mid"
      when 17..20 then "high"
      when 21.. then "blackjack"
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    case
      when card1 == "ace" && card2 == "ace" then "P"
      when self.card_range(card1, card2) == "blackjack"
        if !["ace", "jack", "king", "queen", "ten"].include?(dealer_card)
          "W"
        else
          "S"
        end
      when self.card_range(card1, card2) == "high" then "S"
      when self.card_range(card1, card2) == "mid"
        if  self.parse_card(dealer_card) < 7
          "S"
        else
          "H"
        end
      when self.card_range(card1, card2) == "low" then "H"
    end
  end
end
